python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -320,6 +327,122 @@ def get_downstream_partitions_for_partitions(
         return downstream_partitions_def.empty_subset().with_partition_keys(set(matching_keys))
 
 
+class StaticPartitionMapping(PartitionMapping):
+    """
+    Define an explicit correspondence between two StaticPartitionDefinitions
+
+    Args:
+        mapping (Dict[str, str | Collection[str]]): The single or multi-valued
+            correspondence from upstream keys to downstream keys.
+    """
+
+    def __init__(
+        self,
+        mapping: Dict[str, Union[str, Collection[str]]],
+    ):
+        check.dict_param(
+            mapping, "mapping", key_type=str, value_type=(str, collections.abc.Collection)
+        )
+
+        # cache forward and reverse mappings
+        self._mapping = collections.defaultdict(set)
+        for upstream_key, downstream_keys in mapping.items():
+            self._mapping[upstream_key] = (
+                {downstream_keys} if isinstance(downstream_keys, str) else downstream_keys
+            )
+
+        self._inverse_mapping = collections.defaultdict(set)
+        for upstream_key, downstream_keys in self._mapping.items():
+            for downstream_key in downstream_keys:
+                self._inverse_mapping[downstream_key].add(upstream_key)
+
+        @functools.cache
+        def _check_upstream(upstream_partitions_def: StaticPartitionsDefinition):
+            """
+            validate that the mapping from upstream to downstream is only defined on upstream keys
+            """
+            check.inst_param(
+                upstream_partitions_def,
+                "upstream_partitions_def",
+                StaticPartitionsDefinition,
+                "StaticPartitionMapping can only be defined between two StaticPartitionsDefinitions",
+            )
+            upstream_keys = upstream_partitions_def.get_partition_keys()
+            extra_keys = set(self._mapping.keys()).difference(upstream_keys)
+            if extra_keys:
+                raise ValueError(
+                    f"mapping source partitions not in the upstream partitions definition: {extra_keys}"
+                )
+
+        self._check_upstream = _check_upstream
+
+        @functools.cache
+        def _check_downstream(downstream_partitions_def: StaticPartitionsDefinition):
```

Comment:
the caching should cause this to only be run once. Since constructing the mapping and inverse mapping are also O(N), thought the tradeoff is worth it.

I just like being upfront if the mapping does not accord with the partitions instead of silently not doing anything